### PR TITLE
feat: create register with password hash and login end points (TBC-27)

### PR DESCRIPTION
**Pull request by ohnomalone**
### PR Checklist

Please check for the following requirements:

- [x] I have reviewed this PR's changes at least once
- [ ] The squashed commit message explains exactly what you changed, added or deleted

#### Changes

- register end point:
  - requires `firstName`, `lastName`, `email` (formatted `text@text.com`) and `password` (minimum of 8 characters)
  - custom messaging if they are not supplied or correctly formatted
  - checks if email already is in use if trying to create a duplicate account
- login end point:
  - checks if email already exists then
  - checks if password matches stored password

##### Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

#### What is the current behavior?

- [TBC-27](https://travelingbookclub.atlassian.net/browse/tbc-27)
- [ ] Another developer ran through the acceptance criteria locally (bugs fixes and features)
- [x] I ran through the acceptance criteria locally (bugs fixes and features)
- [ ] I have added tests for meeting the acceptance criteria (bugs fixes and features)

Screenshot 2023-06-07 at 7 31 01 AM
Screenshot 2023-06-07 at 7 32 08 AM

#### Footnotes
- things to consider is the path. I don't think we need the path `users/register` and `users/login`, I think it can just be `/register` and `/login`

[TBC-27]: https://travelingbookclub.atlassian.net/browse/TBC-27?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: backend/controllers/Auth.js
Comment (ohnomalone) [inline backend/controllers/Auth.js:1]: removed this file for use of `authControllers`

File: backend/controllers/authController.js [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+
+const handleErrors = (err) => {
+  if (err.code === 11000) {
+    return { email: 'Email is already registered' };
+  }
+return  Object.keys(err.errors).reduce((acc, key) => {
+    acc[key] = err.errors[key].message;
+  return acc;
+}, {});
+}
```
Comment (ohnomalone) [inline backend/controllers/authController.js:11]: this method checks error codes and errors to build out custom messages for them. Uses reduce to build out object for multiple errors.

File: backend/controllers/authController.js [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+      errors,
+    });
+  }
+}
+
+module.exports.login_post = async (req, res) => {
+  const { email, password } = req.body;
+  try {
+    const user = await User.findOne({ email });
+    if (!user) {
```
Comment (ohnomalone) [inline backend/controllers/authController.js:40]: first we check if there is a user, if not we send a message back there isn't one.

File: backend/controllers/authController.js [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+      return res.status(404).json({
+        message: "User not found",
+      });
+    }
+    const passwordMatch = await user.matchPassword(password);
```
Comment (ohnomalone) [inline backend/controllers/authController.js:45]: If there is a user, we then check if the passed password matches the hashed password using this method that lives on the user schema.

File: backend/controllers/authController.js [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+      return res.status(200).json({
+        userId: user._id,
+        firstName: user.firstName,
+        lastName: user.lastName,
+        email: user.email,
+      });
+    } else {
+      return res.status(401).json({
+        password: "Invalid password",
+      });
```
Comment (ohnomalone) [inline backend/controllers/authController.js:56]: sends back a message if the password doesn't match.

File: backend/models/User.js [hunk 5]
```diff
@@ -1,19 +1,39 @@
 const mongoose = require("mongoose");
-const User = new mongoose.Schema({
+const bcrypt = require('bcrypt');
+
+const UserSchema = new mongoose.Schema({
   firstName: {
     type: String,
-    required: true,
+    required: [true, "Please enter your first name"],
```
Comment (ohnomalone) [inline backend/models/User.js:7]: this sends back a a message if this requirement isn't met.

File: backend/models/User.js [hunk 6]
```diff
@@ -1,19 +1,39 @@
...
+    lowercase: true,
```
Comment (ohnomalone) [inline backend/models/User.js:8]: `lowercase` saves everything as lowercase - makes managing the database easier as well as using text on the front end more consistent.

File: backend/models/User.js [hunk 7]
```diff
@@ -1,19 +1,39 @@
...
+    minlength: [8, "Minimum password length is 8 characters"],
   },
 });
 
-module.exports = mongoose.model("user", User);
+UserSchema.pre('save', async function(next) {
+  const salt = await bcrypt.genSalt();
+  this.password = await bcrypt.hash(this.password, salt);
+  next();
+});
```
Comment (ohnomalone) [inline backend/models/User.js:33]: before the save on the user, the password is hashed. `getSalt` isn't passed a number, by default its 10 and that seems like a pretty secure place to be for it.

File: backend/models/User.js [hunk 8]
```diff
@@ -1,19 +1,39 @@
...
+
+UserSchema.methods.matchPassword = async function(password) {
+  return await bcrypt.compare(password, this.password);
+}
```
Comment (ohnomalone) [inline backend/models/User.js:37]: Here is the method that checks if the give password matches that stored in the database.

File: backend/routes/authRoutes.js
Comment (ohnomalone) [inline backend/routes/authRoutes.js:1]: This routes file will hold our the routes, separating them from the logic that behind it to make it easier to manage.

**Review by watermalon3 — APPROVED** [review]
Amazing JOB!

**Pull request merged**